REGRESSION(289946@main): Web content process crashes when creating attributed string on range containing anchor element with invalid URL

aprotyas opened the pull request:
#### c22ccfdb2bc75cf44985e5a297b74bd01293dbd3
<pre>
REGRESSION(289946@main): Web content process crashes when creating attributed string on range containing anchor element with invalid URL
<a href="https://bugs.webkit.org/show_bug.cgi?id=291335">https://bugs.webkit.org/show_bug.cgi?id=291335</a>
<a href="https://rdar.apple.com/148899219">rdar://148899219</a>

Reviewed by Alex Christensen.

After 289946@main, we return `nil` for `URL::createNSURL()` if the URL
is invalid. As we unconditionally add this result to an attribute dict
under `WebCore::updateAttributes()`, we can now hit an ObjC exception
since we cannot add `nil` items to a NSMutableDictionary.

To fix this crash, we null check the result of `createNSURL()` before we
try to add it to the dictionary.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(updateAttributes):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/anchor-element-with-invalid-href.html: Added.
* Tools/TestWebKitAPI/Tests/WebKit/mac/AttributedSubstringForProposedRange.mm: Renamed from Tools/TestWebKitAPI/Tests/WebKit/mac/AttributedSubstringForProposedRangeWithImage.mm.
(TestWebKitAPI::didFinishNavigation):
(TestWebKitAPI::processDidCrash):
(TestWebKitAPI::invalidMessageFunction):
(TestWebKitAPI::TEST(WebKit, AttributedSubstringForProposedRangeWithImage)):
(TestWebKitAPI::TEST(WebKit, AttributedSubstringForProposedRangeWithAnchorWithInvalidHref)):

Canonical link: <a href="https://commits.webkit.org/293499@main">https://commits.webkit.org/293499@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4741e7da414c90109e887a531970fc5d334d109b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99111 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18754 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8999 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104231 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/49692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101151 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19043 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27191 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/75430 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/49692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102116 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/14469 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89485 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55794 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14259 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7457 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/49068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7545 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106595 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26220 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/84390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26583 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85685 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83903 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21279 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/28566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6238 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/19933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26161 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/31342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25982 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/29295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27548 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2613,8 +2613,8 @@ static void updateAttributes(const Node* node, const RenderStyle& style, OptionS
         [attributes removeObjectForKey:NSLinkAttributeName];
-    else
-        [attributes setObject:linkURL.createNSURL().get() forKey:NSLinkAttributeName];
+    else if (RetainPtr linkNSURL = linkURL.createNSURL())
```
We may want to remove the NSLinkAttributeName key if this is null.  So this could be:
if (RetainPtr linkNSURL = linkURL.createNSURL())
    [attributes setObject:linkNSURL.get() forKey:NSLinkAttributeName];
else
    [attributes setObject:linkURL.createNSURL().get() forKey:NSLinkAttributeName];

aprotyas replied:
Hmm I think I see what you mean, you mean to say we should `removeObjectForKey:NSLinkAttributeName` when the URL is invalid too, right?

```cpp
if (RetainPtr linkNSURL = linkURL.createNSURL(); linkURL.isEmpty() || !linkNSURL)
    [attributes removeObjectForKey:NSLinkAttributeName];
else
    [attributes setObject:linkNSURL.get() forKey:NSLinkAttributeName
```

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKit/mac/AttributedSubstringForProposedRange.mm:
```diff
@@ -82,4 +82,32 @@ static void invalidMessageFunction(WKStringRef messageName)
+    Util::run(&didFinishLoad);
+
+    [webView.platformView() attributedSubstringForProposedRange:NSMakeRange(0, 3) completionHandler: ^(NSAttributedString *attrString, NSRange actualRange) {
+        didFinishTest = true;
```
Can we check the value of the NSAttributedString here?

aprotyas replied:
Sure could, though I don't think we care about the contents too much? (i.e. what about it would we want to check? The actual string/length?)

aprotyas replied:
Added a check for the string content

aprotyas committed:
REGRESSION(289946@main): Web content process crashes when creating attributed string on range containing anchor element with invalid URL
https://bugs.webkit.org/show_bug.cgi?id=291335
rdar://148899219

Reviewed by Alex Christensen.

After 289946@main, we return `nil` for `URL::createNSURL()` if the URL
is invalid. As we unconditionally add this result to an attribute dict
under `WebCore::updateAttributes()`, we can now hit an ObjC exception
since we cannot add `nil` items to a NSMutableDictionary.

To fix this crash, we null check the result of `createNSURL()` before we
try to add it to the dictionary.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(updateAttributes):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/anchor-element-with-invalid-href.html: Added.
* Tools/TestWebKitAPI/Tests/WebKit/mac/AttributedSubstringForProposedRange.mm: Renamed from Tools/TestWebKitAPI/Tests/WebKit/mac/AttributedSubstringForProposedRangeWithImage.mm.
(TestWebKitAPI::didFinishNavigation):
(TestWebKitAPI::processDidCrash):
(TestWebKitAPI::invalidMessageFunction):
(TestWebKitAPI::TEST(WebKit, AttributedSubstringForProposedRangeWithImage)):
(TestWebKitAPI::TEST(WebKit, AttributedSubstringForProposedRangeWithAnchorWithInvalidHref)):

Canonical link: https://commits.webkit.org/293499@main
